src/Infrastructure/ServiceBasedPlugin.php

```diff
@@ -0,0 +1,459 @@
+
+	/**
+	 * Instantiate a Theme object.
+	 *
+	 * @param bool                  $enable_filters    Optional. Whether to
+	 *                                                 enable filtering of the
+	 *                                                 injector configuration.
+	 * @param Injector|null         $injector          Optional. Injector
+	 *                                                 instance to use.
+	 * @param ServiceContainer|null $service_container Optional. Service
+	 *                                                 container instance to
+	 *                                                 use.
+	 */
+	public function __construct(
+		$enable_filters = true,
+		Injector $injector = null,
+		ServiceContainer $service_container = null
+	) {
+		/*
+		 * We use what is commonly referred to as a "poka-yoke" here.
+		 *
+		 * We need an injector and a container. We make them injectable so that
+		 * we can easily provide overrides for testing, but we also make them
+		 * optional and provide default implementations for easy regular usage.
+		 */
+
+		$this->enable_filters = $enable_filters;
+
+		$this->injector = null !== $injector
+			? $injector
+			: new Injector\SimpleInjector();
+
+		$this->injector = $this->configure_injector( $this->injector );
+
+		$this->service_container = null !== $service_container
+			? $service_container
+			: new ServiceContainer\SimpleServiceContainer();
+	}
+
+	/**
+	 * Activate the plugin.
+	 *
+	 * @param bool $network_wide Whether the activation was done network-wide.
+	 * @return void
+	 */
+	public function activate( $network_wide ) {
+		$this->register_services();
+
+		foreach ( $this->service_container as $service ) {
+			if ( $service instanceof Activateable ) {
+				$service->activate( $network_wide );
+			}
+		}
+
+		\flush_rewrite_rules();
+	}
+
+	/**
+	 * Deactivate the plugin.
+	 *
+	 * @param bool $network_wide Whether the deactivation was done network-wide.
+	 * @return void
+	 */
+	public function deactivate( $network_wide ) {
+		$this->register_services();
+
+		foreach ( $this->service_container as $service ) {
+			if ( $service instanceof Deactivateable ) {
+				$service->deactivate( $network_wide );
+			}
+		}
+
+		\flush_rewrite_rules();
+	}
+
+	/**
+	 * Register the plugin with the WordPress system.
+	 *
+	 * @return void
+	 * @throws InvalidService If a service is not valid.
+	 */
+	public function register() {
+		\add_action(
+			static::REGISTRATION_ACTION,
+			[ $this, 'register_services' ]
+		);
+	}
+
+	/**
+	 * Register the individual services of this plugin.
+	 *
+	 * @throws InvalidService If a service is not valid.
+	 *
+	 * @return void
+	 */
+	public function register_services() {
+		// Bail early so we don't instantiate services twice.
+		if ( count( $this->service_container ) > 0 ) {
+			return;
+		}
+
+		// Add the injector as the very first service.
+		$this->service_container->put(
+			static::SERVICE_PREFIX . static::INJECTOR_ID,
+			$this->injector
+		);
```

Comment:
The injector is always added into the service container first.
Right now, the injector would be directly accessible via:
```
AmpWpPluginFactory::create()->get_container()->get( 'injector' );
```

The injector can then be used directly to instantiate stuff without needing to care about the dependencies.

As an example, to instantiate the `SiteHealth` service, you would just do:
```
$site_health = $injector->make( SiteHealth::class );
```
Even though it does have a dependency ( `MonitorCssTransientCaching` ), we don't need to bother with it.